osu.Game/Online/Chat/Message.cs

```diff
@@ -59,19 +60,28 @@ public Message()
         /// <remarks>The <see cref="Link"/>s' <see cref="Link.Index"/> and <see cref="Link.Length"/>s are according to <see cref="DisplayContent"/></remarks>
         public List<Link> Links;
 
+        private static long constructionOrderStatic;
+        private readonly long constructionOrder;
+
         public Message(long? id)
         {
             Id = id;
+
+            constructionOrder = Interlocked.Increment(ref constructionOrderStatic);
         }
 
         public int CompareTo(Message other)
         {
-            if (!Id.HasValue)
-                return other.Id.HasValue ? 1 : Timestamp.CompareTo(other.Timestamp);
-            if (!other.Id.HasValue)
-                return -1;
+            if (Id.HasValue && other.Id.HasValue)
+                return Id.Value.CompareTo(other.Id.Value);
```

Comment:
Can't help but notice that messages with IDs used to use timestamp for comparison in the old code, but now are using IDs. In theory this should be equivalent, but I think this may subtly reorder messages in edge cases, as the timestamp looks to be [generated on the php side](https://github.com/ppy/osu-web/blob/571a3645e9358e61198b24d5fb9c4ea2ce22a60c/app/Models/Chat/Channel.php#L460) when the message is received, while the message ID will presumably be assigned by mysql-side autoincrement. So in theory they may not always be the same given concurrent message submissions. Not sure if actual issue.

Reply:
Yep. I think this is more correct now though, as it will match across the board.